packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -2104,16 +2105,32 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
     return decoration!.hoverColor ?? themeData.inputDecorationTheme.hoverColor ?? themeData.hoverColor;
   }
 
-  Color _getDefaultIconColor(ThemeData themeData) {
-    if (!decoration!.enabled && !isFocused)
-      return themeData.disabledColor;
+  Color _getIconColor(ThemeData themeData) {
```

Comment:
This will not work since themeData is not in the scope of _resolveIconColor, we could use currying to get around this if that would be acceptable solution

```dart
  Color Function(Set<MaterialState>) _resolveIconColor(ThemeData themeData) => (Set<MaterialState> states) {
    if (states.contains(MaterialState.disabled) && !states.contains(MaterialState.focused))
      return themeData.disabledColor;

    if (states.contains(MaterialState.focused))
          return themeData.colorScheme.primary;

    switch (themeData.brightness) {
    case Brightness.dark:
      return Colors.white70;
    case Brightness.light:
      return Colors.black45;
    }
  };

  Color _getIconColor(ThemeData themeData) {
    return MaterialStateProperty.resolveAs(themeData.inputDecorationTheme.iconColor, materialState)
      ?? MaterialStateProperty.resolveWith(_resolveIconColor(themeData)).resolve(materialState);
  }
```

Reply:
Good point and that's pretty slick but it may not help with the "pretty complicated" issue :-).  Factoring out the resolve function makes it a little more obvious that we're just return the first MaterialStateProperty that resolves to null. So:

```dart
Color _getIconColor(ThemeData themeData) {
  Color _resolveIconColor(Set<MaterialState> states) {
    if (states.contains(MaterialState.disabled) && !states.contains(MaterialState.focused))
      return themeData.disabledColor;

    if (states.contains(MaterialState.focused))
          return themeData.colorScheme.primary;

    switch (themeData.brightness) {
    case Brightness.dark:
      return Colors.white70;
    case Brightness.light:
      return Colors.black45;
    }
  }
  return MaterialStateProperty.resolveAs(themeData.inputDecorationTheme.iconColor, materialState)
    ?? MaterialStateProperty.resolveWith(_resolveIconColor).resolve(materialState);
}
```

Reply:
I think for this case your approach is better, will modify